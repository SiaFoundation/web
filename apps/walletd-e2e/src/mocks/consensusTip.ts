import { ConsensusTipResponse } from '@siafoundation/react-walletd'
import { cloneDeep } from '@technically/lodash'
import { Page } from 'playwright'

const data: ConsensusTipResponse = {
  height: 61676,
  id: 'bid:00000010d5da9002b9640d920d9eb9f7502c5c3b2a796ecf800a103920bea96f',
}

export async function mockApiConsensusTipState({ page }: { page: Page }) {
  const json = cloneDeep(data)
  await page.route('**/api/consensus/tipstate*', async (route) => {
    await route.fulfill({ json })
  })
  return json
}
