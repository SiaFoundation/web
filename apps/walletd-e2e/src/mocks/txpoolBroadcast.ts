import { Page } from 'playwright'

export async function mockApiTxpoolBroadcast({ page }: { page: Page }) {
  await page.route(`**/api/txpool/broadcast*`, async (route) => {
    await route.fulfill()
  })
}
