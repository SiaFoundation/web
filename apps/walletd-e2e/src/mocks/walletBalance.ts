import { WalletBalanceResponse } from '@siafoundation/react-walletd'
import { toHastings } from '@siafoundation/units'
import { cloneDeep } from '@technically/lodash'
import { Page } from 'playwright'

const data: WalletBalanceResponse = {
  siacoins: toHastings('100').toString(),
  immatureSiacoins: toHastings('10').toString(),
  siafunds: 10,
}

export async function mockApiWalletBalance({
  page,
  walletId,
  response,
}: {
  page: Page
  walletId: string
  response?: WalletBalanceResponse
}) {
  const json = response || cloneDeep(data)
  await page.route(`**/api/wallets/${walletId}/balance*`, async (route) => {
    await route.fulfill({ json })
  })
  return json
}
