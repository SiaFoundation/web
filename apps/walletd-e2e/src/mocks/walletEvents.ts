import { WalletEvent } from '@siafoundation/react-walletd'
import { cloneDeep } from '@technically/lodash'
import { Page } from 'playwright'

const data: WalletEvent[] = []

export async function mockApiWalletEvents({
  page,
  walletId,
}: {
  page: Page
  walletId: string
}) {
  const json = cloneDeep(data)
  await page.route(`**/api/wallets/${walletId}/events*`, async (route) => {
    await route.fulfill({ json })
  })
  return json
}
