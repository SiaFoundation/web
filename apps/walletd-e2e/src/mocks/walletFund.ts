import { WalletFundResponse } from '@siafoundation/react-walletd'
import { cloneDeep } from '@technically/lodash'
import { Page } from 'playwright'

const data: WalletFundResponse = {
  transaction: {
    siacoinInputs: [
      {
        parentID:
          'scoid:aa3e781330c9b3991e0141807df1327fadf114ca6c37acb9e58004f942d91dfb',
        unlockConditions: {
          timelock: 0,
          publicKeys: null,
          signaturesRequired: 0,
        },
      },
      {
        parentID:
          'scoid:32e430158591b4073a6834e9f4c4b67162e348844f569f4e472896bb72efb724',
        unlockConditions: {
          timelock: 0,
          publicKeys: null,
          signaturesRequired: 0,
        },
      },
    ],
    siacoinOutputs: [
      {
        value: '1000000000000000000000000',
        address:
          'addr:90c6057cdd2463eca61f83796e83152dbba28b6cb9a74831a043833051ec9f422726bfff2ee8',
      },
      {
        value: '97984280000000000000000000',
        address:
          'addr:f2dbf56b5b0c698d7fbf43f646c76169d84e597e8b37fada97348beeecaa812d400ac4ce7981',
      },
    ],
    minerFees: ['3930000000000000000000'],
  },
  toSign: [
    'h:aa3e781330c9b3991e0141807df1327fadf114ca6c37acb9e58004f942d91dfb',
    'h:32e430158591b4073a6834e9f4c4b67162e348844f569f4e472896bb72efb724',
  ],
  dependsOn: null,
}

export async function mockApiWalletFund({
  page,
  walletId,
  response,
}: {
  page: Page
  walletId: string
  response?: WalletFundResponse
}) {
  const json = cloneDeep(response || data)
  await page.route(`**/api/wallets/${walletId}/fund*`, async (route) => {
    await route.fulfill({ json })
  })
  return json
}
