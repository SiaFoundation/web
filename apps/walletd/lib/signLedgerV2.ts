import {
  Result,
  SiafundElement,
  SiacoinElement,
  V2Transaction,
} from '@siafoundation/types'
import { LedgerDevice } from '../contexts/ledger/types'
import { AddressData } from '../contexts/addresses/types'
import { getSDK } from '@siafoundation/sdk'
import { getAddressKeyIndex } from './signV2'

export async function signTransactionLedgerV2({
  device,
  transaction,
  addresses,
  siacoinOutputs,
  siafundOutputs,
}: {
  device: LedgerDevice
  transaction: V2Transaction
  addresses: AddressData[]
  siacoinOutputs: SiacoinElement[]
  siafundOutputs: SiafundElement[]
}): Promise<Result<{ transaction: V2Transaction }>> {
  const base64ToHex = (data: string): string => {
    const buffer: Buffer = Buffer.from(data, 'base64')
    return buffer.toString('hex').slice(0, 128)
  }

  const signatures = new Map<number, string>()
  for (const input of transaction.siacoinInputs ?? []) {
    const indexResponse = getAddressKeyIndex({
      address: input.parent.siacoinOutput.address,
      addresses,
    })

    if ('error' in indexResponse) {
      return { error: indexResponse.error }
    }

    const { index } = indexResponse
    if (!signatures.has(index)) {
      const signResult = await signTransactionIndex({
        device,
        transaction,
        keyIndex: index,
      })

      if ('error' in signResult) {
        return { error: signResult.error }
      }
      signatures.set(index, base64ToHex(signResult.signature))
    }

    input.satisfiedPolicy.signatures = [signatures.get(index)]
  }

  for (const input of transaction.siafundInputs ?? []) {
    const indexResponse = getAddressKeyIndex({
      address: input.parent.siafundOutput.address,
      addresses,
    })

    if ('error' in indexResponse) {
      return { error: indexResponse.error }
    }

    const { index } = indexResponse
    if (!signatures.has(index)) {
      const signResult = await signTransactionIndex({
        device,
        transaction,
        keyIndex: index,
      })

      if ('error' in signResult) {
        return { error: signResult.error }
      }
      signatures.set(index, base64ToHex(signResult.signature))
    }

    input.satisfiedPolicy.signatures = [signatures.get(index)]
  }

  return { transaction }
}

async function signTransactionIndex({
  device,
  transaction,
  keyIndex,
}: {
  device: LedgerDevice
  transaction: V2Transaction
  keyIndex: number
}): Promise<Result<{ signature: string }>> {
  const { encodedTransaction, error } =
    getSDK().wallet.encodeV2Transaction(transaction)
  if (error) {
    return { error }
  }
  const encodedTransactionBuffer = Buffer.from(encodedTransaction, 'utf-8')

  try {
    const signature = await device.sia.signV2Transaction(
      encodedTransactionBuffer,
      0,
      keyIndex,
      0
    )
    return { signature }
  } catch (e) {
    return { error: e.message }
  }
}
