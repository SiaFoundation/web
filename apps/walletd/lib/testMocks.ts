import Sia from '@siacentral/ledgerjs-sia'
import { LedgerDevice } from '../contexts/ledger/types'
import { transformAddressesResponse } from '../contexts/addresses/dataset'
import { Wallet, WalletAddressesResponse } from '@siafoundation/walletd-types'

export function getMockDevice() {
  return {
    type: 'HID',
    sia: {
      transport: {},
      getVersion: jest.fn(() => '0.4.5'),
      signTransaction: jest
        .fn()
        .mockReturnValueOnce(
          'Xt1EJckLmWXU+7HHHDN9bRV5KRuLdC4YY01LzaAMF269QH4hWV8zFkY3kCWs65svhb9HhA1Ix1MRGvhN9orBDpAA'
        )
        .mockReturnValueOnce(
          'fvmSaRzlO/n2L5tsT32e82kWqHnIjQJ8cqjWOc37TtlK6p/vIiOG+TO98HfvbgObTOYVqlKMtUyxTOjGb3bfCpAA'
        ),
    } as unknown as Sia,
    transport: {
      forget: jest.fn(),
      deviceModel: {
        productName: 'Ledger Nano S',
      },
      _disconnectEmitted: false,
    },
  } as LedgerDevice
}

export function getMockAddresses({
  newWallet,
  walletAddressesResponse,
}: {
  newWallet: Wallet
  walletAddressesResponse: WalletAddressesResponse
}) {
  return transformAddressesResponse(
    walletAddressesResponse,
    newWallet.id,
    () => null
  )
}
