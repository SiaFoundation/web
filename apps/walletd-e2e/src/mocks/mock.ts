import { Page } from 'playwright'
import {
  Wallet,
  WalletAddressesResponse,
  WalletBalanceResponse,
  WalletFundResponse,
  WalletOutputsSiacoinResponse,
} from '@siafoundation/react-walletd'
import { mockSiaCentralExchangeRates } from './siaCentralExchangeRates'
import { mockApiSyncerPeers } from './peers'
import { mockApiConsensusTip } from './consensusTipState'
import { mockApiConsensusTipState } from './consensusTip'
import { mockApiConsensusNetwork } from './consensusNetwork'
import { mockApiWallets } from './wallets'
import { mockApiWalletBalance } from './walletBalance'
import { mockApiWalletAddresses } from './walletAddresses'
import { mockApiWalletEvents } from './walletEvents'
import { mockApiWalletTxpool } from './walletTxpool'
import { mockApiWalletOutputsSiacoin } from './walletOutputsSiacoin'
import { mockApiWalletOutputsSiafund } from './walletOutputsSiafund'
import { mockApiWalletFund } from './walletFund'
import { mockApiTxpoolBroadcast } from './txpoolBroadcast'
import { mockApiWalletRelease } from './walletRelease'

export async function mockApiDefaults({ page }: { page: Page }) {
  await mockSiaCentralExchangeRates({ page })
  await mockApiSyncerPeers({ page })
  await mockApiConsensusTip({ page })
  await mockApiConsensusTipState({ page })
  await mockApiConsensusNetwork({ page })
  await mockApiTxpoolBroadcast({ page })
  const wallets = await mockApiWallets({ page })
  for (const wallet of wallets) {
    await mockApiWallet({ page, wallet })
  }
}

export async function mockApiWallet({
  page,
  wallet,
  responses = {},
}: {
  page: Page
  wallet: Wallet
  responses?: {
    balance?: WalletBalanceResponse
    outputsSiacoin?: WalletOutputsSiacoinResponse
    fund?: WalletFundResponse
    addresses?: WalletAddressesResponse
  }
}) {
  await mockApiWalletBalance({
    page,
    walletId: wallet.id,
    response: responses.balance,
  })
  await mockApiWalletAddresses({
    page,
    walletId: wallet.id,
    response: responses.addresses,
  })
  await mockApiWalletEvents({ page, walletId: wallet.id })
  await mockApiWalletTxpool({ page, walletId: wallet.id })
  await mockApiWalletOutputsSiacoin({
    page,
    walletId: wallet.id,
    response: responses.outputsSiacoin,
  })
  await mockApiWalletOutputsSiafund({ page, walletId: wallet.id })
  await mockApiWalletFund({
    page,
    walletId: wallet.id,
    response: responses.fund,
  })
  await mockApiWalletRelease({
    page,
    walletId: wallet.id,
  })
}
