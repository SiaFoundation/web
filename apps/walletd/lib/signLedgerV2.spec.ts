import { initSDK } from '@siafoundation/sdk'
import { signTransactionLedgerV2 } from './signLedgerV2'
import { getMockAddresses, getMockDevice } from './testMocks'
import { getMockScenarioSeedWallet } from './mocks/mockSeedWallet'

beforeEach(async () => {
  await initSDK()
})

describe('signLedgerV2', () => {
  describe('siacoin', () => {
    it('builds and signs valid transaction', async () => {
      const device = getMockDevice()
      const mocks = getMockScenarioSeedWallet()
      expect(
        await signTransactionLedgerV2({
          device,
          transaction: mocks.walletConstructV2Response.transaction,
          addresses: getMockAddresses(mocks),
        }),
      ).toMatchSnapshot()
    })

    it('errors when a public keys address is missing', async () => {
      const device = getMockDevice()
      const mocks = getMockScenarioSeedWallet()
      expect(
        await signTransactionLedgerV2({
          device,
          transaction: mocks.walletConstructV2Response.transaction,
          addresses: [
            {
              id: 'id',
              walletId: 'id',
              address: 'address not in addresses',
              metadata: {
                index: 5,
              },
            },
          ],
        }),
      ).toEqual({
        error:
          'Missing address 90c6057cdd2463eca61f83796e83152dbba28b6cb9a74831a043833051ec9f422726bfff2ee8',
      })
    })

    it('errors when an address is missing its index', async () => {
      const device = getMockDevice()
      const mocks = getMockScenarioSeedWallet()
      expect(
        await signTransactionLedgerV2({
          device,
          transaction: mocks.walletConstructV2Response.transaction,
          addresses: [
            {
              id: 'id',
              walletId: 'id',
              address:
                mocks.walletOutputsSiacoinResponse.outputs[1].siacoinOutput
                  .address,
              metadata: {},
            },
          ],
        }),
      ).toEqual({
        error:
          'Missing address 90c6057cdd2463eca61f83796e83152dbba28b6cb9a74831a043833051ec9f422726bfff2ee8',
      })
    })
  })
})
