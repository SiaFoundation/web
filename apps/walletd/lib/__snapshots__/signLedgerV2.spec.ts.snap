// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signLedgerV2 siacoin builds and signs valid transaction 1`] = `
Object {
  "transaction": Object {
    "minerFee": "3930000000000000000000",
    "siacoinInputs": Array [
      Object {
        "parent": Object {
          "id": "aa3e781330c9b3991e0141807df1327fadf114ca6c37acb9e58004f942d91dfb",
          "leafIndex": 304248,
          "maturityHeight": 0,
          "merkleProof": Array [
            "0a7a4c392f78899e3c38c5cd9e6a673b2c7afec97930af539af9c8e20209aa78",
            "a1e074dc48634a234b7366a0d7ab19cd05e3e698e1d44bf07e24d75ae0c65b3c",
          ],
          "siacoinOutput": Object {
            "address": "90c6057cdd2463eca61f83796e83152dbba28b6cb9a74831a043833051ec9f422726bfff2ee8",
            "value": "1000000000000000000000000",
          },
        },
        "satisfiedPolicy": Object {
          "policy": Object {
            "policy": Object {
              "publicKeys": Array [
                "ed25519:ee122b2169bdae5776b55609e384e0c58372cd5c529d4edc9b9918b26f8e5535",
              ],
              "signaturesRequired": 1,
              "timelock": 0,
            },
            "type": "uc",
          },
          "signatures": Array [
            "5edd4425c90b9965d4fbb1c71c337d6d1579291b8b742e18634d4bcda00c176ebd407e21595f331646379025aceb9b2f85bf47840d48c753111af84df68ac10e",
          ],
        },
      },
      Object {
        "parent": Object {
          "id": "32e430158591b4073a6834e9f4c4b67162e348844f569f4e472896bb72efb724",
          "leafIndex": 305723,
          "maturityHeight": 0,
          "merkleProof": Array [
            "8c02aeec48de589ce497ebe72fb8b527cfe022ef513fcfdc56745c84832f00ec",
            "1bf63b9959e60272fd7a48a8cecd4120a852c0e14557ea27ccad6ea2071e70b3",
          ],
          "siacoinOutput": Object {
            "address": "f2dbf56b5b0c698d7fbf43f646c76169d84e597e8b37fada97348beeecaa812d400ac4ce7981",
            "value": "97984280000000000000000000",
          },
        },
        "satisfiedPolicy": Object {
          "policy": Object {
            "policy": Object {
              "publicKeys": Array [
                "ed25519:ee122b2169bdae5776b55609e384e0c58372cd5c529d4edc9b9918b26f8e5535",
              ],
              "signaturesRequired": 1,
              "timelock": 0,
            },
            "type": "uc",
          },
          "signatures": Array [
            "7ef992691ce53bf9f62f9b6c4f7d9ef36916a879c88d027c72a8d639cdfb4ed94aea9fef222386f933bdf077ef6e039b4ce615aa528cb54cb14ce8c66f76df0a",
          ],
        },
      },
    ],
    "siacoinOutputs": Array [
      Object {
        "address": "90c6057cdd2463eca61f83796e83152dbba28b6cb9a74831a043833051ec9f422726bfff2ee8",
        "value": "1000000000000000000000000",
      },
      Object {
        "address": "f2dbf56b5b0c698d7fbf43f646c76169d84e597e8b37fada97348beeecaa812d400ac4ce7981",
        "value": "97984280000000000000000000",
      },
    ],
  },
}
`;
